### ISSUE TITLE
How to find Average Precision (AP) for each class or category in YOLO-based models such as YOLOv5, YOLOv8, YOLOv11, or YOLOv12?

### ISSUE THREAD

**Issue by Shahriar-RL** [ultralytics/yolov5#13576]
### Search before asking

- [ ] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

During training, I’d like to automatically display the Average Precision (AP), Precision (P), and Recall (R) for each detection category in every training epoch. What modifications to the code would enable this functionality?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Shahriar-RL, thank you for your interest in YOLOv5 🚀! Please visit our ⭐️ [Tutorials](https://docs.ultralytics.com/yolov5/) to get started, where you can find quickstart guides for simple tasks like [Custom Data Training](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data/) all the way to advanced concepts like [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution/).

If this is a 🐛 Bug Report, please provide a **minimum reproducible example** to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

## Requirements

[**Python>=3.8.0**](https://www.python.org/) with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) installed including [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). To get started:
```bash
git clone https://github.com/ultralytics/yolov5  # clone
cd yolov5
pip install -r requirements.txt  # install
```

## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml"><img src="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml/badge.svg" alt="YOLOv5 CI"></a>

If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are currently passing. CI tests verify correct operation of YOLOv5 [training](https://github.com/ultralytics/yolov5/blob/master/train.py), [validation](https://github.com/ultralytics/yolov5/blob/master/val.py), [inference](https://github.com/ultralytics/yolov5/blob/master/detect.py), [export](https://github.com/ultralytics/yolov5/blob/master/export.py) and [benchmarks](https://github.com/ultralytics/yolov5/blob/master/benchmarks.py) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response 🤖. An Ultralytics engineer will review your question and assist you here soon!

**Comment by pderrenger** [comment]
Hi @Shahriar-RL,

YOLOv5 already calculates per-class metrics during validation, but they're not displayed in the default training output. You can access these metrics in several ways:

1. Check the validation results in `runs/train/exp*/results.txt` which contains per-class AP values

2. In your training loop, you can modify `test.py` or use the callback system to print these values during training

3. For real-time visualization, integrate with experiment tracking tools like ClearML or W&B that automatically chart per-class metrics

If you need more granular control, the metrics are calculated in `utils/metrics.py` where you can add custom logging for your specific needs.

Would you like a specific code example for any of these approaches?

**Comment by vittorio-prodomo** [comment]
> Hi [@Shahriar-RL](https://github.com/Shahriar-RL),
> 
> YOLOv5 already calculates per-class metrics during validation, but they're not displayed in the default training output. You can access these metrics in several ways:
> 
> 1. Check the validation results in `runs/train/exp*/results.txt` which contains per-class AP values
> 2. In your training loop, you can modify `test.py` or use the callback system to print these values during training
> 3. For real-time visualization, integrate with experiment tracking tools like ClearML or W&B that automatically chart per-class metrics
> 
> If you need more granular control, the metrics are calculated in `utils/metrics.py` where you can add custom logging for your specific needs.
> 
> Would you like a specific code example for any of these approaches?

Hi!
So basically, are you saying that ClearML and W&B will already show out-of-the-box per-class AP metrics **during** training? And what about Tensorboard? Can you point me in where should I work (it's fine if I have to modify some code) to enable visualization of per-class AP metrics also in tensorboard logs?

**Comment by pderrenger** [comment]
Hi @vittorio-prodomo — W&B and ClearML will chart per‑class metrics during training as soon as you log per‑class scalars; YOLOv5’s built‑ins log class‑wise PR curves and the confusion matrix, but AP/P/R per‑class scalars aren’t emitted by default. For TensorBoard, expose the per‑class arrays from ap_per_class in val.py and log them where tb_writer is used in train.py, e.g.:

# val.py (inside run(), after computing p, r, ap, f1, ap_class)
per_class = dict(P=p, R=r, mAP50=ap[:, 0], mAP50_95=ap.mean(1))
return (mp, mr, map50, map, loss, speed, …), per_class

# train.py (after validation call)
(results, per_class) = val.run(...)
if tb_writer:
    for i, name in names.items():
        tb_writer.add_scalar(f'metrics/P/{name}', per_class['P'][i], epoch)
        tb_writer.add_scalar(f'metrics/R/{name}', per_class['R'][i], epoch)
        tb_writer.add_scalar(f'metrics/mAP50/{name}', per_class['mAP50'][i], epoch)
        tb_writer.add_scalar(f'metrics/mAP50-95/{name}', per_class['mAP50_95'][i], epoch)

Swap the add_scalar lines for wandb.log(...) or your ClearML logger to get the same plots there. For an overview of TB logging with YOLO, see Visualize YOLO Training Metrics with TensorBoard, and refer to the Ultralytics Docs for details.
